File: python/560.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        n = len(nums)
        presum = [0] * (n+1)
        hashmap = {0: 1}
        res = 0
        for i in range(1, n+1):
            next_sum = presum[i-1] + nums[i-1]
            presum[i] =  next_sum
            if (next_sum - k) in hashmap: res += hashmap[next_sum - k]
            if next_sum not in hashmap: hashmap[next_sum] = 0
            hashmap[next_sum] += 1
        
        return res

File: cpp/560.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int n = nums.size();
        int presum[n+1];
        unordered_map<int, int> map;
        map[0] = 1;
        presum[0] = 0;
        int count = 0;
        for (int i = 1; i <= n; ++i) {
            int next_sum = presum[i-1] + nums[i-1];
            presum[i] = next_sum;
            if (map.find(next_sum - k) != map.end()) count+=map[next_sum - k];
            if (map.find(next_sum) == map.end()) map[next_sum] = 0;
            ++map[next_sum];
        }
        return count;
    }
};


